File: Platforms/Common/Reign_Video_D3D11_Component/States/RasterizerState.cpp
#include "pch.h"
#include "RasterizerState.h"

namespace Reign_Video_D3D11_Component
{
	#pragma region Constructors
	RasterizerStateError RasterizerStateCom::Init(VideoCom^ video, RasterizerStateDescCom^ desc)
	{
		null();
		this->video = video;

		ID3D11RasterizerState* stateTEMP;
		if (FAILED(video->device->CreateRasterizerState(desc->desc, &stateTEMP)))
		{
			return RasterizerStateError::RasterizerState;
		}
		state = stateTEMP;

		return RasterizerStateError::None;
	}

	RasterizerStateCom::~RasterizerStateCom()
	{
		if (state) state->Release();
		null();
	}

	void RasterizerStateCom::null()
	{
		state = 0;
	}
	#pragma endregion

	#pragma region Methods
	void RasterizerStateCom::Enable()
	{
		video->deviceContext->RSSetState(state);
	}
	#pragma endregion
}

File: Platforms/Common/Reign.Video.D3D11/States/RasterizerState.cs
using System;
using Reign.Core;
using Reign_Video_D3D11_Component;

namespace Reign.Video.D3D11
{
	public class RasterizerState : DisposableResource, IRasterizerState
	{
		#region Properties
		private RasterizerStateCom com;
		#endregion

		#region Constructors
		public RasterizerState(IDisposableResource parent, IRasterizerStateDesc desc)
		: base(parent)
		{
			try
			{
				var video = parent.FindParentOrSelfWithException<Video>();

				com = new RasterizerStateCom();
				var error = com.Init(video.com, ((RasterizerStateDesc)desc).com);

				if (error == RasterizerStateError.RasterizerState) Debug.ThrowError("RasterizerState", "Failed to create RasterizerState");
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (com != null)
			{
				com.Dispose();
				com = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public void Enable()
		{
			com.Enable();
		}
		#endregion
	}
}